### PR TITLE
Use `__getstate__()` and `__setstate__()`

### PR DESCRIPTION
**Pull request by jan-janssen**
Based on an offline discussion I had with @pmrv , we would like to propose the use of `__getstate__()` and `__setstate__()` to allow storing any kind of object using `h5io`. This functionality is implemented as optional feature which can be enabled using the `use_state` parameter. It allows `h5io` to store any python class which can be pickled, still rather than using binary strings like they are used by pickle, the objects are stored as standard python types as they are already supported by `h5io`. This has the advantage that individual parts of an object can be loaded from the HDF5 file rather than being required to reload the full object.

### CODE REVIEW

**Comment by jan-janssen** [comment]
This functionality was added in Python 3.11: `Changed in version 3.11: Added the default implementation of the __getstate__() method in the object class.` https://docs.python.org/3/library/pickle.html#object.__getstate__

**Comment by larsoner** [comment]
Seems reasonable to me. I guess it should only be enabled for Python 3.11+ then?

**Comment by jan-janssen** [comment]
> Seems reasonable to me. I guess it should only be enabled for Python 3.11+ then?

I disabled the tests for python 3.11. Or do you mean adding an error message when the option is used on Python 3.11? Do you have any preference what kind of error should be raised? Maybe an `OSError` ?

**Comment by larsoner** [comment]
I think `RuntimeError` probably fits better than `OSError` because the Python version isn't really the operating system's fault (the [description of OSError is](https://docs.python.org/3/library/exceptions.html#OSError) "This exception is raised when a system function returns a system-related error").

**Comment by larsoner** [comment]
... and yes I think this should be raised for Python < 3.11

**Comment by jan-janssen** [comment]
> ... and yes I think this should be raised for Python < 3.11

In principle developers could implement the `__getstate__()` method already before. With Python 3.11 it was introduced as default to the object class, so there is a `__getstate__()` for any object. Still I agree it is better to fail explicitly.

**Comment by larsoner** [comment]
Yeah it's a bit tricky because then there are some objects for which the `__get/set_state__` would work with any Python version. But if we don't raise an error someone could `write_hdf5` on Python 3.11+ and then someone on 3.10 or earlier could try to read it and get a weird error. Simplest just to raise an error as is done here but if you have use cases where `__get/setstate__` does exist we could relax it somehow perhaps.

Looks like just some minor non-auto-fixable style stuff to resolve

**Comment by jan-janssen** [comment]
> Looks like just some minor non-auto-fixable style stuff to resolve

I fixed the Docstrings in the tests - all tests pass now. In terms of the restrictions to Python 3.11 that is fine for me. I think it is the more clean solution.

**Comment by jan-janssen** [comment]
@larsoner I hope you had a relaxing holiday break. I was wondering if you have a chance to look at this pull request again and in case you decide to merge it, release a new `h5io` version. I would really like to use this new functionality in the https://github.com/pyiron package.

File: h5io/_h5io.py [hunk 1]
```diff
@@ -333,6 +348,56 @@ def _triage_write(
                 cleanup_data.append({title: (rootname, key, value)})
                 return
 
+        if use_state:
+            class_type = str(type(value)).split("'")[1]
```
Comment (larsoner) [inline h5io/_h5io.py:352]: I would have thought this maybe more robust/standard than relying on the `__repr__` being `<class 'whatever'>`?
```suggestion
            class_type = value.__class__.__name__
```

**Review by larsoner — COMMENTED** [review]
Other than one other tiny thought, LGTM!

File: h5io/_h5io.py (hunk 1 above)
Comment (jan-janssen) [reply h5io/_h5io.py]: I used `value.__class__.__module__ + "." + value.__class__.__name__` to get both the module class and its name.

**Comment by jan-janssen** [comment]
> Other than one other tiny thought, LGTM!

@larsoner Thanks for your feedback - I updated the pull request accordingly.

**Comment by larsoner** [comment]
Thanks @jan-janssen !

**Pull request merged**